client/pluginmanager/drivermanager/instance.go

```diff
@@ -0,0 +1,443 @@
+	go func() {
+		i.fingerprint()
+		wg.Done()
+	}()
+
+	// Start event handler
+	wg.Add(1)
+	go func() {
+		i.handleEvents()
+		wg.Done()
+	}()
+
+	// Do a final cleanup
+	wg.Wait()
+	i.cleanup()
+}
+
+// dispense is used to dispense a plugin.
+func (i *instanceManager) dispense() (plugin drivers.DriverPlugin, err error) {
+	i.pluginLock.Lock()
+	defer i.pluginLock.Unlock()
+
+	// See if we already have a running instance
+	if i.plugin != nil && !i.plugin.Exited() {
+		return i.driver, nil
+	}
+
+	var pluginInstance loader.PluginInstance
+
+	if reattach, ok := i.fetchReattach(); ok {
+		// Reattach to existing plugin
+		pluginInstance, err = i.loader.Reattach(i.id.Name, i.id.PluginType, reattach)
+	} else {
+		// Get an instance of the plugin
+		pluginInstance, err = i.loader.Dispense(i.id.Name, i.id.PluginType, i.pluginConfig, i.logger)
+	}
+	if err != nil {
+		// Retry as the error just indicates the singleton has exited
+		if err == singleton.SingletonPluginExited {
+			pluginInstance, err = i.loader.Dispense(i.id.Name, i.id.PluginType, i.pluginConfig, i.logger)
+		}
+
+		// If we still have an error there is a real problem
+		if err != nil {
+			return nil, fmt.Errorf("failed to start plugin: %v", err)
+		}
+	}
+
+	// Convert to a fingerprint plugin
+	driver, ok := pluginInstance.Plugin().(drivers.DriverPlugin)
+	if !ok {
+		pluginInstance.Kill()
+		return nil, fmt.Errorf("plugin loaded does not implement the driver interface")
+	}
+
+	// Store the plugin and driver
+	i.plugin = pluginInstance
+	i.driver = driver
+
+	// Store the reattach config
+	if c, ok := pluginInstance.ReattachConfig(); ok {
+		i.storeReattach(c)
+	}
+
+	return driver, nil
+}
+
+// cleanup shutsdown the plugin
+func (i *instanceManager) cleanup() {
+	i.shutdownLock.Lock()
+	i.pluginLock.Lock()
+	defer i.pluginLock.Unlock()
+	defer i.shutdownLock.Unlock()
+
+	if i.plugin != nil && !i.plugin.Exited() {
+		i.plugin.Kill()
+		i.storeReattach(nil)
+	}
+}
+
+// dispenseFingerprintCh dispenses a driver and makes a Fingerprint RPC call
+// to the driver. The fingerprint chan is returned along with the cancel func
+// for the context used in the RPC. This cancel func should always be called
+// when the caller is finished with the channel.
+func (i *instanceManager) dispenseFingerprintCh() (<-chan *drivers.Fingerprint, context.CancelFunc, error) {
+	driver, err := i.dispense()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ctx, cancel := context.WithCancel(i.ctx)
+	fingerCh, err := driver.Fingerprint(ctx)
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+
+	return fingerCh, cancel, nil
+}
+
+// fingerprint is the main loop for fingerprinting.
+func (i *instanceManager) fingerprint() {
+	fpChan, cancel, err := i.dispenseFingerprintCh()
+	if err != nil {
+		i.logger.Error("failed to dispense driver plugin", "error", err)
+	}
+
+	// backoff and retry used if the RPC is closed by the other end
+	var backoff time.Duration
+	var retry int
+	for {
+		time.Sleep(backoff)
```

Comment:
Select on time.After(backoff) and Done (with perhaps an `if backoff > 0` check to skip the select if it's not in a retry loop).